File: go_algorithms/algorithms/ColumnByColumnParallel.go
package algorithms

import (
	"fmt"
	"sync"
	"time"
)

func ColumnByColumnParallel(A, B [][]int) ([][]int, int64) {
	fmt.Println("Método: ColumnByColumnParallel")

	startTime := time.Now().UnixNano()

	rowsA, colsA, colsB := len(A), len(A[0]), len(B[0])

	result := make([][]int, rowsA)
	for i := range result {
		result[i] = make([]int, colsB)
	}

	var wg sync.WaitGroup
	wg.Add(colsA * colsB)

	for j := 0; j < colsB; j++ {
		for i := 0; i < rowsA; i++ {
			go func(row, col int) {
				defer wg.Done()
				sum := 0
				for k := 0; k < colsA; k++ {
					sum += A[i][k] * B[k][j]
				}
				result[row][col] = sum
			}(i, j)
		}
	}

	wg.Wait()

	endTime := time.Now().UnixNano()
	executionTime := endTime - startTime
	return result, executionTime
}

File: python_algorithms/algorithms/ColumnByColumnParallel.py
import numpy as np
import threading
import time


def Multiply(A, B):
        print("Método: ColumnByColumnParallel")

        start_time = time.time_ns()

        rowsA, colsA, colsB = len(A), len(A[0]), len(B[0])

        result = np.zeros((rowsA, colsB), dtype=np.int64)

        def calculate_element(row, col):
            sum_val = 0
            for k in range(colsA):
                sum_val += A[row][k] * B[k][col]
            result[row][col] = sum_val

        threads = []
        for j in range(colsB):
            for i in range(rowsA):
                thread = threading.Thread(target=calculate_element, args=(i, j))
                thread.start()
                threads.append(thread)

        for thread in threads:
            thread.join()

        end_time = time.time_ns()
        execution_time = end_time - start_time
        return result, execution_time
